deploy/multiple_keys.sql

```diff
@@ -0,0 +1,83 @@
+--
+
+BEGIN;
+
+CREATE TABLE keys(
+   id char(32) NOT NULL,
+   key_id text NOT NULL,
+   public_key text NOT NULL,
+   key_version integer NOT NULL,
+   created_at TIMESTAMP WITHOUT TIME ZONE NOT NULL,
+   expires_at TIMESTAMP WITHOUT TIME ZONE NOT NULL,
+   PRIMARY KEY (id, key_id)
+   );
+
+CREATE OR REPLACE FUNCTION add_key() RETURNS TRIGGER AS $add_key$
+   BEGIN
+     IF NEW.public_key IS NOT NULL THEN
+       INSERT INTO keys
+         (id, key_id, public_key, key_version, created_at,  expires_at) VALUES
+         (NEW.id, 'default', NEW.public_key, NEW.pubkey_version, now(), 'infinity'::timestamp);
+     END IF;
+     RETURN NULL; -- result is ignored since this is an AFTER trigger
+   END;
+$add_key$ LANGUAGE plpgsql;
+
+CREATE OR REPLACE FUNCTION update_key() RETURNS TRIGGER AS $update_key$
+   BEGIN
+     IF NEW.public_key IS NOT NULL THEN
+       UPDATE keys SET public_key = NEW.public_key, key_version = NEW.pubkey_version
+         WHERE id = NEW.id AND key_id = 'default';
+     ELSE
+       DELETE FROM keys WHERE id = NEW.id AND key_id = 'default';
+     END IF;
+     RETURN NULL; -- result is ignored since this is an AFTER trigger
+   END;
+$update_key$ LANGUAGE plpgsql;
+
+
+CREATE OR REPLACE FUNCTION delete_keys() RETURNS TRIGGER AS $delete_keys$
+  BEGIN
+     DELETE FROM keys WHERE id = OLD.id;
+     RETURN NULL;
+  END;
+$delete_keys$ LANGUAGE plpgsql;
+
+DROP TRIGGER IF EXISTS add_key ON clients;
+DROP TRIGGER IF EXISTS update_key ON clients;
+DROP TRIGGER IF EXISTS delete_keys ON clients;
+
+DROP TRIGGER IF EXISTS add_key ON users;
+DROP TRIGGER IF EXISTS update_key ON users;
+DROP TRIGGER IF EXISTS delete_keys ON users;
+
+CREATE TRIGGER add_key AFTER INSERT ON clients FOR EACH ROW EXECUTE PROCEDURE add_key();
+CREATE TRIGGER update_key AFTER UPDATE ON clients FOR EACH ROW EXECUTE PROCEDURE update_key();
+CREATE TRIGGER delete_keys AFTER DELETE ON clients FOR EACH ROW EXECUTE PROCEDURE delete_keys();
+
+CREATE TRIGGER add_key AFTER INSERT ON users FOR EACH ROW EXECUTE PROCEDURE add_key();
+CREATE TRIGGER update_key AFTER UPDATE ON users FOR EACH ROW EXECUTE PROCEDURE update_key();
+CREATE TRIGGER delete_keys AFTER DELETE ON users FOR EACH ROW EXECUTE PROCEDURE delete_keys();
+
+--
+-- This should simplify lookup in the authorization check function.
+-- Authz id is added to make this a one stop query.
+-- 
+-- This needs some benchmarking on real data sizes; this may trigger sequential scans
+-- when we do something like
+-- select id,org_id,name,type from keys_by_name where name ='wei' AND (org_id='global' OR org_id='181131e4e896ec930f5ca3f36b490c1f');; 
+-- 
+CREATE OR REPLACE view keys_by_name AS 
+SELECT id, org_id, name, authz_id, 'client' AS type, key_id, keys.public_key, key_version FROM clients INNER JOIN keys USING (id) 
+UNION ALL
+SELECT id, 'global', username AS name, authz_id, 'user' AS type, key_id, keys.public_key, key_version FROM users INNER JOIN keys USING (id);
+
```

Comment:
I was thinking of UUID conflict, but I wasn't 100% certain that the ids in this case were UUIDs. Given that my worry of id conflicts would be overblown, as you said.